src/OrleansRuntime/MembershipService/MembershipOracleData.cs

```diff
@@ -271,6 +300,76 @@ private void NotifyLocalSubscribers(SiloAddress siloAddress, SiloStatus newStatu
             }
         }
 
+        private struct UpdateFaultCombo : IComparable<UpdateFaultCombo>
+        {
+            public int UpdateZone;
+            public int FaultZone;
+        
+            public int CompareTo(UpdateFaultCombo other)
+            {
+ 	             var i = UpdateZone.CompareTo(other.UpdateZone);
+                 return (i != 0) ? i : FaultZone.CompareTo(other.FaultZone);
+           }
+        }
+    
+
+        // deterministic function for designating the silos that should act as gateways
+        private List<SiloAddress> DetermineMultiClusterGateways()
+        {
+            Debug.Assert(!string.IsNullOrEmpty(GlobalServiceId)); // call only if this is a multi cluster
+
+            var candidates = localTableCopyOnlyActive.Keys.ToList();
+
+            // sort deterministically so everyone ends up with the same result
+            candidates.Sort((a, b) =>  
+            {
+                var addressdiff = a.Endpoint.Address.ToString().CompareTo(b.Endpoint.Address.ToString());
+                if (addressdiff != 0) return addressdiff;
+                return a.Endpoint.Port.CompareTo(b.Endpoint.Port);
+            });
+
+            // take all the active silos if their count does not exceed the desired number of gateways
+            if (candidates.Count <= NumMultiClusterGatewaysPerCluster)
+                return candidates;
```

Comment:
Should this be named MaxNumMultiClusterGatewaysPerCluster if it could actually be less than that?